File: lua/queen-attack/queen-attack.lua
---Class represents a queen on the board. It has a coordinate within the
---board and a method to check if it can attack another piece (another
---queen, in this case).
---@class Queen
---@field row integer
---@field column integer
---@field can_attack fun(Queen):boolean

---Lowest value on the axis that a coordinator can have.
local BOARD_START = 0
---Highest value on the axis that a coordinator can have.
local BOARD_END = 7

local new, can_attack
local shares_column, shares_diagonal, shares_row

---Creates a new instance of Queen, with the column and row given by
---the argument table.
---@param pos {row: integer, column: integer}
---@return Queen
function new(pos)
    assert(pos.row ~= nil, "missing row")
    assert(pos.column ~= nil, "missing column")

    assert(pos.row >= BOARD_START, "row underflow")
    assert(pos.column >= BOARD_START, "column underflow")

    assert(pos.row <= BOARD_END, "row overflow")
    assert(pos.column <= BOARD_END, "column overflow")

    local self = { row = pos.row, column = pos.column }

    ---Returns whether or not this queen can attack the given queen "other".
    ---@param other Queen
    ---@return boolean
    function self.can_attack(other)
        return can_attack(self, other)
    end

    return self
end

---Returns whether or not the first queen can attack the second.
---@param self Queen
---@param other Queen
---@return boolean
function can_attack(self, other)
    return shares_column(self, other) or
        shares_row(self, other) or
        shares_diagonal(self, other)
end

---Returns whether the first queen and the second queen are on the same
---column of the board
---@param self Queen
---@param other Queen
---@return boolean
function shares_column(self, other)
    return self.column == other.column
end

---Returns whether or not the first and second queens are on the same
---diagonal of the board.
---@param self Queen
---@param other Queen
---@return boolean
function shares_diagonal(self, other)
    local cols_distance = math.abs(self.column - other.column)
    local rows_distance = math.abs(self.row - other.row)

    return cols_distance == rows_distance
end

---Returns whether or not the first and second queens are on the same
---line of the board.
---@param self Queen
---@param other Queen
---@return boolean
function shares_row(self, other)
    return self.row == other.row
end

return new

File: php/queen-attack/QueenAttack.php
<?php

declare(strict_types=1);

/**
 * Validates if a given coordinate pair represents a valid position on a standard 8x8 chessboard.
 *
 * A valid position must have both X (rank) and Y (file) coordinates between 0 and 7, inclusive.
 *
 * @param int $xCoordinate The X-coordinate (rank) of the queen's position.
 * @param int $yCoordinate The Y-coordinate (file) of the queen's position.
 * @return bool True if the position is valid.
 * @throws InvalidArgumentException If the coordinates are out of bounds (less than 0 or greater than 7).
 */
function placeQueen(int $xCoordinate, int $yCoordinate): bool
{
    if ($xCoordinate > 7 || $yCoordinate > 7) {
        throw new InvalidArgumentException('The position must be on a standard size chess board.');
    }

    if ($xCoordinate < 0 || $yCoordinate < 0) {
        throw new InvalidArgumentException('The rank and file numbers must be positive.');
    }

    return true;
}

/**
 * Determines if two queens placed on a chessboard can attack each other.
 *
 * Queens can attack if they are on the same rank (row), same file (column), or same diagonal.
 * This function utilizes helper functions to check each of these conditions.
 *
 * @param array<int> $whiteQueen An array representing the coordinates [x, y] of the white queen.
 * @param array<int> $blackQueen An array representing the coordinates [x, y] of the black queen.
 * @return bool True if either queen can attack the other, false otherwise.
 *
 * @example
 * canAttack([0, 0], [0, 7]); // Returns true (same rank)
 * canAttack([0, 0], [7, 0]); // Returns true (same file)
 * canAttack([0, 0], [6, 6]); // Returns true (same diagonal)
 * canAttack([0, 0], [1, 2]); // Returns false
 */
function canAttack(array $whiteQueen, array $blackQueen): bool
{
    return array_any(
        ["isInSameRank", "isInSameFile", "isInSameDiagonal"],
        fn($cond) => $cond($whiteQueen, $blackQueen)
    );
}

/**
 * Checks if two queens are in the same rank (row).
 *
 * @param array<int> $whiteQueen The coordinates [x, y] of the white queen.
 * @param array<int> $blackQueen The coordinates [x, y] of the black queen.
 * @return bool True if their X-coordinates are identical, false otherwise.
 */
function isInSameRank(array $whiteQueen, array $blackQueen): bool
{
    return $whiteQueen[0] === $blackQueen[0];
}

/**
 * Checks if two queens are in the same file (column).
 *
 * @param array<int> $whiteQueen The coordinates [x, y] of the white queen.
 * @param array<int> $blackQueen The coordinates [x, y] of the black queen.
 * @return bool True if their Y-coordinates are identical, false otherwise.
 */
function isInSameFile(array $whiteQueen, array $blackQueen): bool
{
    return $whiteQueen[1] === $blackQueen[1];
}

/**
 * Checks if two queens are in the same diagonal.
 *
 * Queens are on the same diagonal if the absolute difference between their X-coordinates
 * is equal to the absolute difference between their Y-coordinates.
 *
 * @param array<int> $whiteQueen The coordinates [x, y] of the white queen.
 * @param array<int> $blackQueen The coordinates [x, y] of the black queen.
 * @return bool True if they are on the same diagonal, false otherwise.
 */
function isInSameDiagonal(array $whiteQueen, array $blackQueen): bool
{
    $rankDiff = abs($whiteQueen[0] - $blackQueen[0]);
    $fileDiff = abs($whiteQueen[1] - $blackQueen[1]);

    return $rankDiff === $fileDiff;
}